### PR TITLE
backends subpackage

### PR DESCRIPTION
**Pull request by niboshi**
This PR introduces `chainer.backends` subpackage to include various backend modules.

### CODE REVIEW

**Comment by niboshi** [comment]
I think this PR may be easier to review commit-by-commit, rather than by merged diff.

**Comment by niboshi** [comment]
It's not dependent, but please merge #3850 and #3851 first. Otherwise conflict resolution would be cumbersome.

**Comment by niboshi** [comment]
Rebased.

File: chainer/__init__.py [hunk 1]
```diff
@@ -4,8 +4,8 @@
 import warnings
 
 from chainer import _version
+from chainer import backends  # NOQA
```
Comment (kmaehashi) [inline chainer/__init__.py]: I think it is better to fix following two lines too.

```
from chainer.cuda import should_use_cudnn  # NOQA
from chainer.cuda import should_use_cudnn_tensor_core  # NOQA
```

File: chainer/__init__.py [hunk 2]
```diff
@@ -64,6 +64,7 @@
 
 
 from chainer import _environment_check
+from chainer.backends import cuda
```
Comment (kmaehashi) [inline chainer/__init__.py]: Do we still need this even though `cuda.py` exists?

File: chainer/__init__.py (hunk 1 above)
Comment (niboshi) [reply chainer/__init__.py]: Thank you. Fixed.

**Comment by kmaehashi** [comment]
LGTM, could you please resolve conflict?

**Comment by niboshi** [comment]
Rebased.

**Comment by kmaehashi** [comment]
Thank you! I'll merge after Travis pass.

**Comment by kmaehashi** [comment]
LGTM!

**Comment by kmaehashi** [comment]
I'll backport this after merging #4019.

**Pull request merged**